### ISSUE TITLE
Some nodes turn black or disappear when using shadow bake

### ISSUE THREAD

**Issue by SkanerSoft** [godotengine/godot#45574]
**Godot version:**
3.2.4 RC 1

**OS/device including version:**
Huawei | MTK-MT6762R | Android 10 | GLES 2

**Issue description:**
Some nodes turn black after export to android or disappear..
At the same time, the console displays the following errors:
```
E 0:00:07.809   _display_error_with_code: SceneShaderGLES2: Fragment shader compilation failed:
Compile failed.
ERROR: 0:716: 'array type' : requires language version 300
ERROR: main() function is missing.
2 compilation errors. No code generated.
  <C++ Source>  drivers/gles2/shader_gles2.cpp:130 @ _display_error_with_code()
E 0:00:07.810   get_current_version: Method failed. Returning: __null
  <C++ Source>  drivers/gles2/shader_gles2.cpp:377 @ get_current_version()
E 0:00:07.810   bind: Condition "!version" is true. Returned: false
  <C++ Source>  drivers/gles2/shader_gles2.cpp:89 @ bind()
E 0:00:07.810   _get_uniform: Condition "!version" is true. Returned: -1
  <C++ Source>  drivers/gles2/shader_gles2.h:267 @ _get_uniform()
```

The error occurs when using the new CPULightmapper.
BVH does not affect the error.

Sometimes the nodes are not displayed at all, although they are present on the stage.

**Steps to reproduce:**
1. Create a Project
2. Make a scene
3. Make objects on scene (export glTF from Blender)
3.1 Unwrap, set a "use baking"
4. Add a CPULightmapper
5. Bake
6. Add dynamic objects (not "use baking")

Run this project on PC - OK
Run this project on Android - ERROR

**Screenshots**
on PC
![Image 017](https://user-images.githubusercontent.com/13809475/106352828-2c9f1780-62f7-11eb-9f0f-b63943292698.jpg)

on Android
![Screenshot_20210130_122440_ru test error](https://user-images.githubusercontent.com/13809475/106352829-2d37ae00-62f7-11eb-8b42-83e2fa07006f.jpg)

**Minimal reproduction project:**
https://drive.google.com/file/d/1SV-Wn-Swe2tqlETwcoAo3Xy3P8jyVdR9/view?usp=sharing

**Comment by clayjohn** [comment]
Cc @JFonS I'm not sure how the new lightmapper works in GLES2. But it looks like it is relying on texture arrays, which are only supported by certain hardware (almost all desktop devices, but few Android devices). We need to have an appropriate fallback for when texture arrays aren't supported.

**Comment by JFonS** [comment]
That's weird, the GLES2 shader code should not have any usage of texture arrays at all...

Since the error refers to "array types" I assume it means GLSL's support for arrays of basic types (as in vec2[4], for example).
The only place I could find such arrays in the scene shader code is in the lightmap's capture.

If my theory is correct, the issue should be gone if you disable capture before baking (in the BakedLighmap node, under "Capture"). On top of that, the issue should also be present in previous versions of the engine, since the new lightmapper doesn't change the capture code at all.

@SkanerSoft If you could test this and report your results that would be great :)

**Comment by SkanerSoft** [comment]
I did as you asked, but the result hasn't changed.
On startup, I get the shader code (the console cuts it off), and errors. Same as before

Here are the screenshots to make sure I did everything right. The result has not changed, everything is still black and some objects have disappeared.

![Image 002](https://user-images.githubusercontent.com/13809475/106362172-735e3300-6332-11eb-8cad-734533098a99.jpg)
![Image 004](https://user-images.githubusercontent.com/13809475/106362173-73f6c980-6332-11eb-8647-78c5c799d3fc.jpg)
![Image 005](https://user-images.githubusercontent.com/13809475/106362175-748f6000-6332-11eb-8b37-816599b18d80.jpg)

**Comment by JFonS** [comment]
@SkanerSoft Just to double-check, did you re-bake the lightmap after disabling capture?

I still think the issue is the same as in #45584, as @Calinou pointed out. In theory disabling capture should fix things, but I'm probably missing something. I will take a closer look.

**Comment by SkanerSoft** [comment]
@JFonS Of course, after removing "Capture", I reloaded the editor and baked the map again.

**Comment by SkanerSoft** [comment]
Okay, I understand why the scene is black. When baking the shadows, an image of 4938x4938 pixels is obtained.
The phone simply cannot download (or process) it.
I can manually reduce the image with a slight loss in quality. Can I lower it in the settings? I set everything to "low", but the size of the shadow maps is still very large.

As for the disappearing objects, you probably need to do something about that too :)

**Comment by JFonS** [comment]
@SkanerSoft ok, thanks for the testing :)

The lightmap size is defined by the "Lightmap Size Hint" property of your `Mesh` resources. If you are unwrapping meshes at import time (setting the "Light Baking" import option to "Gen Lightmaps") this value is automatically computed. You can adjust the generated lightmap size by changing the "Lightmap Texel SIze" import option, which specifies the approximate size of each texel in word units. 

As a recap, there are 3 separate issues, I think:

1. GSLS arrays are not supported in some devices (and apparently HTML5), so lightmap captures don't work.
2. Capture data is kept in the `BakedLightmapData` resources, even after re-baking with capture disabled. This should be easy to fix.
3. Some devices can't handle large textures. We should probably document this can happen and how to reduce the lightmap size.

**Comment by SkanerSoft** [comment]
Thank you for the clarification! I will work with texels.
As for the disappearance of objects, of course, the solution with disabling capture works, but it breaks the atmosphere. With GLES3, the game is terribly slow on the phone.

Dynamic objects are now always light)
![Image 001](https://user-images.githubusercontent.com/13809475/106364494-82e47880-6340-11eb-97e9-896a5893d327.jpg)

**Comment by Zireael07** [comment]
> With GLES3, the game is terribly slow on the phone.

GLES3 is almost never a good choice for mobile games (their GLES3 drivers suck). Pick GLES2 if you target mobile.

**Comment by SkanerSoft** [comment]
> Pick GLES2 if you target mobile

Of course) It's a pity that the capture doesn't work. Hopefully this will be fixed. Otherwise, all my level design doesn't make sense.

**Comment by Calinou** [comment]
>      Of course) It's a pity that the capture doesn't work. Hopefully this will be fixed. Otherwise, all my level design doesn't make sense.

It might be argued that brighter enemies are a feature to help with visibility :slightly_smiling_face:

**Comment by SkanerSoft** [comment]
This is very sad because they have a style. I have a bug.
But the joke is funny)

**Comment by JFonS** [comment]
Should be fixed by #45622.

**Comment by akien-mga** [comment]
Fixed by #45752.
